### PR TITLE
[1334] integrate nightwatch

### PR DESCRIPTION
**Pull request by RVKen**

### CODE REVIEW

File: test/e2e/nightwatch.json [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    "type" : "mocha",
+    "options" : {
+      "ui" : "bdd",
+      "reporter" : "list"
+    }
+  },
+
+  "selenium" : {
+    "start_process" : true,
+    "server_path" : "test/e2e/bin/selenium-server-standalone-3.4.0.jar",
```
Comment (fehguy) [inline test/e2e/nightwatch.json]: I don't think we should put the jar files in the repo.  Can you pull it down during the build?

File: test/e2e/nightwatch.json [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    "log_path" : "",
+    "host" : "127.0.0.1",
+    "port" : 4444,
+    "cli_args" : {
+      "webdriver.chrome.driver" : "test/e2e/bin/chromedriver",
```
Comment (fehguy) [inline test/e2e/nightwatch.json]: Same as above, we should not put binaries in the repo.  Let's see if there are alternatives to doing so

File: test/e2e/scenarios/test.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+    afterEach(function (client, done) {
+      done()
+    })
+
+    beforeEach(function (client, done) {
+      done()
+    })
+
+    it('uses TDD to run the Google simple test', function (client) {
```
Comment (fehguy) [inline test/e2e/scenarios/test.js]: Can we remove the google test and replace it with a test against the specification you added at `1.json`?

File: test/e2e/specs/1.json [hunk 4]
```diff
@@ -0,0 +1 @@
...roperties":{"id":{"type":"integer","format":"int64"},"username":{"type":"string"},"firstName":{"type":"string"},"lastName":{"type":"string"},"email":{"type":"string"},"password":{"type":"string"},"phone":{"type":"string"},"userStatus":{"type":"integer","format":"int32","description":"User Status"}},"xml":{"name":"User"}},"Tag":{"type":"object","properties":{"id":{"type":"integer","format":"int64"},"name":{"type":"string"}},"xml":{"name":"Tag"}},"Pet":{"type":"object","required":["name","photoUrls"],"properties":{"id":{"type":"integer","format":"int64"},"category":{"$ref":"#/definitions/Category"},"name":{"type":"string","example":"doggie"},"photoUrls":{"type":"array","xml":{"name":"photoUrl","wrapped":true},"items":{"type":"string"}},"tags":{"type":"array","xml":{"name":"tag","wrapped":true},"items":{"$ref":"#/definitions/Tag"}},"status":{"type":"string","description":"pet status in the store","enum":["available","pending","sold"]}},"xml":{"name":"Pet"}},"ApiResponse":{"type":"object","properties":{"code":{"type":"integer","format":"int32"},"type":{"type":"string"},"message":{"type":"string"}}}},"externalDocs":{"description":"Find out more about Swagger","url":"http://swagger.io"}}
```
Comment (fehguy) [inline test/e2e/specs/1.json]: Can you please pretty-print this?

File: test/e2e/nightwatch.json (hunk 1 above)
Comment (RVKen) [inline test/e2e/nightwatch.json]: yeah, i also prefer that but couldn't figure it out due to shortage of time. maybe we can write some instructions in readme to download & put the bin files into the desired location.

File: test/e2e/scenarios/test.js (hunk 3 above)
Comment (RVKen) [inline test/e2e/scenarios/test.js]: Sure, this is just for demo on my approach.

File: test/e2e/specs/1.json (hunk 4 above)
Comment (RVKen) [inline test/e2e/specs/1.json]: Right. Should be done in my next PR.

File: test/e2e/nightwatch.json (hunk 1 above)
Comment (ponelat) [inline test/e2e/nightwatch.json]: try this https://www.npmjs.com/package/selenium-server-standalone-jar

Comment (ponelat) [inline test/e2e/nightwatch.json]: or https://www.npmjs.com/package/webdriver-manager for both selenium server and chromedriver bins. Probably preferred

Comment (shockey) [inline test/e2e/nightwatch.json]: +1 for managing test dependencies via npm 😄

File: test/e2e/scenarios/initial-render/test.js [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+      })
+    })
+
+    afterEach(function (client, done) {
+      done()
+    })
+
+    beforeEach(function (client, done) {
+      mainPage = client
+        .url("localhost:3200")
```
Comment (fehguy) [inline test/e2e/scenarios/initial-render/test.js]: at first, i wasn't able to figure out where swagger-ui is rendering from.  But it looks like from here, you're running a webserver serving up swagger-ui here on port 3200?  If that's correct could you please see about having the webserver start up as part of the test (on some arbitrary port?) and hit it from there?  We'll have to also pass on the spec as a JSON object or via URL, which can be hosted by this same webserver.

**Comment by RVKen** [comment]
Yes, right. swagger-ui needs to be run manually for now. Both your concerns should be resolved gracefully.

File: package.json [hunk 6]
```diff
@@ -122,6 +123,7 @@
     "react-test-renderer": "^15.5.4",
     "rimraf": "^2.6.0",
     "sass-loader": "^6.0.2",
+    "selenium-server-standalone-jar": "^3.4.0",
```
Comment (ponelat) [inline package.json]: use a fixed version, if a newer version gets installed, then `server_path` will be incorrect. 
`"selenium-server-standalone-jar": "3.4.0",`

File: package.json [hunk 7]
```diff
@@ -32,48 +32,51 @@
...
     "just-test": "karma start --config karma.conf.js",
-    "just-test-in-node": "mocha --recursive --compilers js:babel-core/register test/core test/components test/bugs test/swagger-ui-dist-package"
+    "just-test-in-node": "mocha --recursive --compilers js:babel-core/register test/core test/components test/bugs test/swagger-ui-dist-package",
+    "e2e": "nightwatch test/e2e/scenarios --config test/e2e/nightwatch.json --verbose",
+    "e2e-initial-render": "nightwatch test/e2e/scenarios --config test/e2e/nightwatch.json --group initial-render",
+    "mock:api": "json-server --watch test/e2e/db.json --port 3204"
   },
   "dependencies": {
     "base64-js": "^1.2.0",
-    "brace": "0.7.0",
```
Comment (shockey) [inline package.json]: **Action required:** Please don't bump production dependencies in unrelated PRs 😄

File: package.json [hunk 8]
```diff
@@ -32,48 +32,51 @@
     "test": "npm run lint-errors && npm run just-test-in-node",
     "test-in-node": "npm run lint-errors && npm run just-test-in-node",
     "just-test": "karma start --config karma.conf.js",
-    "just-test-in-node": "mocha --recursive --compilers js:babel-core/register test/core test/components test/bugs test/swagger-ui-dist-package"
+    "just-test-in-node": "mocha --recursive --compilers js:babel-core/register test/core test/components test/bugs test/swagger-ui-dist-package",
+    "e2e": "nightwatch test/e2e/scenarios --config test/e2e/nightwatch.json --verbose",
```
Comment (shockey) [inline package.json]: would prefer `test-e2e` instead of `e2e`

File: package.json [hunk 9]
```diff
@@ -32,48 +32,51 @@
...
+    "e2e-initial-render": "nightwatch test/e2e/scenarios --config test/e2e/nightwatch.json --group initial-render",
```
Comment (shockey) [inline package.json]: would prefer `test-e2e-initial-render`

File: package.json [hunk 10]
```diff
@@ -32,48 +32,51 @@
...
+    "mock:api": "json-server --watch test/e2e/db.json --port 3204"
```
Comment (shockey) [inline package.json]: i like the `:` in theory, but I'd prefer `mock-api` for consistency

File: dev-helpers/template.ejs [hunk 11]
```diff
@@ -5,7 +5,6 @@
   <meta charset="UTF-8">
```
Comment (shockey) [inline dev-helpers/template.ejs]: `dev-helpers/index.html` needs to be intact, since the hot dev server uses it - unless you've configured the dev server to use this template as well.

File: package.json [hunk 12]
```diff
@@ -18,7 +18,6 @@
     "build-bundle": "webpack --config webpack-dist-bundle.config.js --colors",
     "build-core": "webpack --config webpack-dist.config.js --colors",
     "build-standalone": "webpack --config webpack-dist-standalone.config.js --colors",
-    "predev": "npm install",
```
Comment (shockey) [inline package.json]: Why was this removed?

File: package.json [hunk 13]
```diff
@@ -32,48 +31,51 @@
...
     "test-in-node": "npm run lint-errors && npm run just-test-in-node",
     "just-test": "karma start --config karma.conf.js",
-    "just-test-in-node": "mocha --recursive --compilers js:babel-core/register test/core test/components test/bugs test/swagger-ui-dist-package"
+    "just-test-in-node": "mocha --recursive --compilers js:babel-core/register test/core test/components test/bugs test/swagger-ui-dist-package",
+    "test-e2e": "sleep 3 && nightwatch test/e2e/scenarios --config test/e2e/nightwatch.json",
+    "test-e2e-initial-render": "nightwatch test/e2e/scenarios --config test/e2e/nightwatch.json --group initial-render",
+    "mock-api": "json-server --watch test/e2e/db.json --port 3204",
+    "e2e-flow": "npm-run-all --parallel -r hot-server mock-api test-e2e"
   },
   "dependencies": {
```
Comment (shockey) [inline package.json]: Please revert changes to production dependency versions 😄 

Dev dependencies can be bumped if everything still works.

File: README.md [hunk 14]
```diff
@@ -58,6 +58,13 @@ If you'd like to make modifications to the codebase, run the dev server with: `n
 
 If you'd like to rebuild the `/dist` folder with your codebase changes, run `npm run build`.
 
+
+##### Integrate Tests
```
Comment (shockey) [inline README.md]: `Integrate` should be `Integration`.

File: README.md [hunk 15]
```diff
@@ -58,6 +58,13 @@ If you'd like to make modifications to the codebase, run the dev server with: `n
...
+
+You will need JDK of version 7 or higher as instructed here
+http://nightwatchjs.org/gettingstarted#selenium-server-setup
```
Comment (shockey) [inline README.md]: 1) Can we reasonably expect a system to have JDK 7+ installed?

2) If we want these integration tests to automatically run in CI (do we?), we need to make sure we're provisioning Travis correctly. @webron?

File: dev-helpers/template.ejs [hunk 16]
```diff
@@ -67,37 +66,37 @@
 
 <div id="swagger-ui"></div>
 
-<script src="./swagger-ui-bundle.js"> </script>
```
Comment (shockey) [inline dev-helpers/template.ejs]: I see this was removed; where is the bundle and standalone preset coming from now that this is gone?

File: package.json (hunk 12 above)
Comment (clever2000) [inline package.json]: Reverted

File: dev-helpers/template.ejs (hunk 16 above)
Comment (clever2000) [inline dev-helpers/template.ejs]: Reverted

File: dev-helpers/template.ejs (hunk 11 above)
Comment (clever2000) [inline dev-helpers/template.ejs]: This was changed to ejs template for my previous approach and I forgot to revert it. Now index.html is back.

Comment (RVKen) [inline dev-helpers/template.ejs]: This was changed to ejs template for my previous approach and I forgot to revert it. Now index.html is back.

File: dev-helpers/template.ejs (hunk 16 above)
Comment (RVKen) [inline dev-helpers/template.ejs]: Reverted

File: package.json (hunk 12 above)
Comment (RVKen) [inline package.json]: Reverted

File: package.json (hunk 13 above)
Comment (RVKen) [inline package.json]: Can you please elaborate on this? I don't think I understand this.

File: README.md (hunk 14 above)
Comment (RVKen) [inline README.md]: Done

File: README.md (hunk 15 above)
Comment (RVKen) [inline README.md]: 1. That is a must for selenium and i think it should be configured manually.

File: package.json [hunk 17]
```diff
@@ -32,48 +32,51 @@
     "test": "npm run lint-errors && npm run just-test-in-node",
     "test-in-node": "npm run lint-errors && npm run just-test-in-node",
     "just-test": "karma start --config karma.conf.js",
-    "just-test-in-node": "mocha --recursive --compilers js:babel-core/register test/core test/components test/bugs test/swagger-ui-dist-package"
+    "just-test-in-node": "mocha --recursive --compilers js:babel-core/register test/core test/components test/bugs test/swagger-ui-dist-package",
+    "test-e2e": "sleep 3 && nightwatch test/e2e/scenarios --config test/e2e/nightwatch.json",
```
Comment (shockey) [inline package.json]: Just wondering - why are we doing `sleep 3` here?

File: package.json [hunk 18]
```diff
@@ -32,48 +32,51 @@
...
+    "test-e2e-initial-render": "nightwatch test/e2e/scenarios/ --config test/e2e/nightwatch.json --group initial-render",
+    "mock-api": "json-server --watch test/e2e/db.json --port 3204",
+    "e2e-flow": "npm-run-all --parallel -r hot-server mock-api test-e2e"
   },
   "dependencies": {
```
Comment (shockey) [inline package.json]: Production dependency version ranges should not be changed in this PR.

File: package.json (hunk 17 above)
Comment (RVKen) [inline package.json]: We're running mock-api and test-e2e simultaneously and it is to ensure `mock-api` is up and running before test-e2e runs.

File: package.json [hunk 19]
```diff
@@ -32,7 +32,11 @@
...
     "test-in-node": "npm run lint-errors && npm run just-test-in-node",
     "just-test": "karma start --config karma.conf.js",
-    "just-test-in-node": "mocha --recursive --compilers js:babel-core/register test/core test/components test/bugs test/swagger-ui-dist-package"
+    "just-test-in-node": "mocha --recursive --compilers js:babel-core/register test/core test/components test/bugs test/swagger-ui-dist-package",
+    "test-e2e": "sleep 3 && nightwatch test/e2e/scenarios/ --config test/e2e/nightwatch.json",
+    "e2e-initial-render": "nightwatch test/e2e/scenarios/ --config test/e2e/nightwatch.json --group initial-render",
+    "mock-api": "json-server --watch test/e2e/db.json --port 3204",
+    "e2e": "npm-run-all --parallel -r hot-server mock-api test-e2e"
   },
   "dependencies": {
```
Comment (shockey) [inline package.json]: The e2e test dependencies should be in `devDependencies`.

**Comment by shockey** [comment]
I only spotted one thing that needs to be changed - the new test stuff still needs to be moved out of `dependencies`, otherwise npm end users will pull down unnecessary e2e assets when installing swagger-ui.

I'd also like to see the tests split into multiple files - it's easier to read, easier to work with (for those of us that work in [text editors with sketchy offscreen buffers](//atom.io/)), and will allow Nightwatch to run more tests in parallel:

> Test concurrency is done at the file level. Each test file will fill a test worker slot. Individual tests/steps in a test file will not run concurrently.
>
> http://nightwatchjs.org/guide/#parallel-running

Without parallel test runs, the e2e test battery takes just over four minutes to complete, which is a bit of a bummer - mostly for me, since I spend a significant amount of time waiting for tests to pass before merging PRs.

**Comment by shockey** [comment]
Merged 🎉 

Thanks @RVKen!
